modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/maven/AbstractMavenClassLoaderModelLoader.java

```diff
@@ -161,14 +176,43 @@ protected ClassLoaderModel createHeavyPackageClassLoaderModel(File artifactFile,
           .collect(toSet());
     }
 
-    loadUrls(artifactFile, classLoaderModelBuilder, bundleDependencies);
+    // TODO MULE-17114 retrieve this data from the json if present, if not then call this
+    populateLocalPackages(artifactFile, classLoaderModelBuilder, bundleDependencies, exportedPackages, exportedResources);
+
     classLoaderModelBuilder.dependingOn(bundleDependencies);
 
     return classLoaderModelBuilder.build();
   }
 
+  protected void populateLocalPackages(File artifactFile, final ArtifactClassLoaderModelBuilder classLoaderModelBuilder,
+                                       Set<BundleDependency> bundleDependencies,
+                                       Set<String> exportedPackages, Set<String> exportedResources) {
+    // This is already filtering out mule-plugin dependencies,
+    // since for this case we explicitly need to consume the exported API from the plugin.
+    final List<URL> dependenciesArtifactsUrls = loadUrls(artifactFile, classLoaderModelBuilder, bundleDependencies);
+    for (URL dependencyArtifactUrl : dependenciesArtifactsUrls) {
+      final URI dependencyArtifactUri;
+      try {
+        dependencyArtifactUri = dependencyArtifactUrl.toURI();
+      } catch (URISyntaxException e) {
+        throw new MuleRuntimeException(e);
+      }
+
+      final JarInfo exploredJar = jarExplorerFactory.get().explore(dependencyArtifactUri);
+
+      final Set<String> localPackages = new HashSet<>(exploredJar.getPackages());
+      localPackages.removeAll(exportedPackages);
```

Comment:
I think that this will only cover one part of what @pablokraan has suggested to change here (original comment was https://github.com/mulesoft/mule/pull/8029#discussion_r301590317):
**You can only ignore packages from the region (the parent first ones) when they are not part of your exported API.** 
I think that system packages have to be ignored here too: **SYSTEM_PACKAGES (ie: org.mule.runtime, com.mulesoft.mule.runtime)** but the part that I'm having hard time to understand is:
**Packages in the closure that are already provided by the region cannot be overridden by the artifact. Otherwise you can get ClassCastExceptions.**. 
I understand this but at the same time this is exactly what we need to do here, in order to avoid loading spring packages from the sharedLibraries (API from owner classloader in region) and use the spring packages from the internal library bundled within the plugin. So, how does this work with the general rule from Kraan's comment?

Reply:
I keep thinking about this and perhaps this is already covered, let say that you have artifact A that exports packageA and packageB as it depends on artifact B, so artifact B exports only packageB. At the same time what you are doing is mark as local package a packageC which comes from introspecting the libraries in artifact A (it depends on a library C).
Now, how about this case, you also have an artifact C in the region (another plugin) which exports a package C. In that case, as artifact A from its API is not requiring to "import" the package C, it will be using its local classes from its own classloader assuming that you will never have to access to those classes from artifact C. right?
If that's the case, I think these changes are fine and you could just dismiss my previous comment.

Reply:
yes, from what we talked with @pablokraan there are some caes where a plugin may be handling objects created by another plugin, and of both plugins have the same local class a `ClassCastException` may occur. I think these cases would be very rare, since there would be a common interface in the region for handling those, but that is not being enforced. We should discuss how those kind of static validations may be enforced in the packager.